Use zero? for comparison (#172)
andycroll/andycroll.com: _posts/2022/2022-10-31-use-zero-for-integer-float-comparison.md

---
title: "Use zero? for comparison of numerics like Integer, Float and BigDecimal"
description: "Other languages don't have this"
layout: article
category: ruby
image:
  base: "2022/use-zero-for-integer-float-comparison"
  alt: "A zero on a sign on the wall"
  credit: "Bernard Hermant"
  source: "https://unsplash.com/photos/rfpSOlH1JlQ"
---

> Ruby often takes a more “readable” approach to its syntax and standard library of included methods than other languages.

Reviewer: Compared to other languages, Ruby prioritises readability when it comes to it syntax and standard library ?

Author: ```suggestion
Compared to other similar languages Ruby often prioritises readability (and joy) when it come to its syntax and methods in its standard library.
```

> Compared to other similar languages Ruby often prioritises readability (and joy) when it come to its syntax and methods in its standard library.

Reviewer: ```suggestion
Compared to other similar languages, Ruby often prioritises readability (and joy) when it comes to its syntax and the methods provided in its Standard Library.
```

> An example of this is the common programming comparison of numeric variable to zero.

Reviewer: This sentence isn't so clear.

Author: ```suggestion
An example of this is some syntactic sugar for comparing a value to zero.
```

> An example of this is some syntactic sugar for comparing a value to zero.

Reviewer: ```suggestion
An example of this is the syntactic sugar used when comparing a value to zero.
```

> See the documentation of the `.zero?` method on [`Integer`](https://ruby-doc.org/core-3.1.2/Integer.html#zero-3F-method), [`Float`](https://ruby-doc.org/core-3.1.2/Float.html#zero-3F-method), [`Numeric`](https://ruby-doc.org/core-3.1.2/Numeric.html#zero-3F-method) and [`BigDecimal`](https://ruby-doc.org/stdlib-3.1.2/libdoc/bigdecimal/rdoc/BigDecimal.html#zero-3F-method).

Reviewer: Hmm, at this point, not clear on where this article is going.

Author: IT'S A JOURNEY.

A short one.

> See the documentation of the `.zero?` method on [`Integer`](https://ruby-doc.org/core-3.1.2/Integer.html#zero-3F-method), [`Float`](https://ruby-doc.org/core-3.1.2/Float.html#zero-3F-method), [`Numeric`](https://ruby-doc.org/core-3.1.2/Numeric.html#zero-3F-method) and [`BigDecimal`](https://ruby-doc.org/stdlib-3.1.2/libdoc/bigdecimal/rdoc/BigDecimal.html#zero-3F-method).

Reviewer: *Take a look at the...

Sounds a little more conversational/fun?

Author: ```suggestion
Take a look at the documentation for the `.zero?` method on [`Integer`](https://ruby-doc.org/core-3.1.2/Integer.html#zero-3F-method), [`Float`](https://ruby-doc.org/core-3.1.2/Float.html#zero-3F-method), [`Numeric`](https://ruby-doc.org/core-3.1.2/Numeric.html#zero-3F-method) and [`BigDecimal`](https://ruby-doc.org/stdlib-3.1.2/libdoc/bigdecimal/rdoc/BigDecimal.html#zero-3F-method).
```

## Instead of...

> ...checking whether a value is equal to zero in a conditional:

Reviewer: *equal to '0'

Author: ```suggestion
...checking whether a value is equal to `0` in a conditional:
```

> ...checking whether a value is zero in a conditional, how other languages might do it:

Reviewer: *which is common in other languages: ?

Author: ```suggestion
...checking whether a value is equal to zero in a conditional:
```

```ruby
if number == 0
  :yes
else
  :no
end
```

## Use...

...the built-in `#zero?` method on all numeric types:

```ruby
if number.zero?
  :yes
else
  :no
end
```

## Why?

> The Ruby-ish syntax in this case is clearer and less error prone.

Reviewer: ```suggestion
The Ruby-ish syntax is clearer and less error prone.
```

> It is very easy to type `number = 0` and thus assign a variable in the conditional rather than compare (although you might use tests to protect yourself from this particular error).

Reviewer: ```suggestion
In the first example, it is very easy to mistakenly type `number = 0` and thus assign a variable in the conditional rather than perform a comparison (although you might use tests to protect yourself from that particular error).
```

## Why not?

> The `== 0` syntax is how _many_ other similar languages perform comparison and is deeply ingrained in my fingers. I still struggle to apply this advice even after 15 years of Ruby programming.

Reviewer: *comparisons

*deeply ingrained for many of us

Author: ```suggestion
The `== 0` syntax is how _many_ other similar languages perform comparisons and is deeply ingrained for many of us. I still struggle to apply this advice even after 15 years of Ruby programming.
```

> This is a habit I still struggle with after 15 years of Ruby programming. The `== 0` syntax is how _many_ other similar languages do comparison and is deeply ingrained in my fingers!

Reviewer: Maybe lead with the issue of how many other languages do it...and then say "Ever after 15 years of Ruby programming..."

Author: ```suggestion
The `== 0` syntax is how _many_ other similar languages perform comparison and is deeply ingrained in my fingers. I still struggle to apply this advice even after 15 years of Ruby programming.
```

> You might get folks pushing back for “performance” reasons.

Reviewer: *some folks

Author: ```suggestion
You might get some folks pushing back for “performance” reasons.
```

> The `== 0` syntax is how _many_ other similar languages perform comparison and is deeply ingrained in my fingers. I still struggle to apply this advice even after 15 years of Ruby programming.

Author: ```suggestion
The `== 0` syntax is how _many_ other similar languages perform comparisons and is deeply ingrained for many of us. I still struggle to apply this advice even after 15 years of Ruby programming.
```

  x.report("0.zero?") { 0.zero? } #=> true

  x.report("1.0 == 0") { 1.0 == 0 } #=> false
  x.report("0.0 == 0") { 0.0 == 0 } #=> true
  x.report("1.0.zero?") { 1.0.zero? } #=> false
  x.report("0.0.zero?") { 0.0.zero? } #=> true
end
```

For `Integer`:

<table>
<tr>
  <th>1 == 0</th>
  <td class="text-right">29.049M <small>(± 0.7%)</small> i/s</td>
</tr>
<tr>
  <th>0 == 0</th>
  <td class="text-right">28.972M <small>(± 0.3%)</small> i/s</td>
</tr>
<tr>
  <th>1.zero?</th>
  <td class="text-right">23.009M <small>(± 1.0%)</small> i/s</td>
</tr>
<tr>
  <th>0.zero?</th>
  <td class="text-right">22.860M <small>(± 1.7%)</small> i/s</td>
</tr>
</table>

For `Float`:

<table>
<tr>
  <th>1.0 == 0</th>
  <td class="text-right">18.768M <small>(± 2.0%)</small> i/s</td>
</tr>
<tr>
  <th>0.0 == 0</th>
  <td class="text-right">19.066M <small>(± 0.5%)</small> i/s</td>
</tr>
<tr>
  <th>1.0.zero?</th>
  <td class="text-right">22.784M <small>(± 0.6%)</small> i/s</td>
</tr>
<tr>
  <th>0.0.zero?</th>
  <td class="text-right">22.841M <small>(± 0.8%)</small> i/s</td>
</tr>
</table>

The benchmark shows that for `Integer`s the `==` syntax is faster, but for `Float`s the `.zero?` syntax wins out.

> However the important thing to note is that in all cases (on my machine) you get about 20 million executions per second so in _your_ code consider the readability over the performance implications!

Reviewer: I think this summary could be said more clearly and remove reference to your machine if possible to make the argument more strongly.

Author: ```suggestion
However the important thing to note is that in all cases you get millions of executions per second so in _your_ code you should emphasise the readability over any perceived performance implications!
```

> However the important thing to note is that in all cases you get millions of executions per second so in _your_ code you should emphasise the readability over any perceived performance implications!

Reviewer: Yes, much better!

```suggestion
However, the important thing to note is that in all cases you still get millions of executions per second, so in _your_ code you should emphasise the readability over any perceived performance implications!
```
